File: core/bitboard.py
"""Bitwise flood-fill BFS over the 9x9 Quoridor board.

Bit index = row * 9 + col. Movement is computed with per-direction
"can move from here" masks derived from the wall sets, matching
core.rules.is_blocked exactly. BFS ignores the opponent pawn.
"""

N = 9
FULL = (1 << (N * N)) - 1


def _bit(c, r):
    return 1 << (r * N + c)


# Static per-row / per-col masks.
_ROW = [sum(_bit(c, r) for c in range(N)) for r in range(N)]
_COL = [sum(_bit(c, r) for r in range(N)) for c in range(N)]

_GOAL_ROW_MASK = {0: _ROW[N - 1], 1: _ROW[0]}

# Source cells from which a step in a given direction is impossible due to the
# board edge (independent of walls).
_EDGE_N = _ROW[N - 1]        # cannot go north from top row
_EDGE_S = _ROW[0]            # cannot go south from bottom row
_EDGE_E = _COL[N - 1]        # cannot go east from right col
_EDGE_W = _COL[0]            # cannot go west from left col


def _can_move_masks(state):
    """Return (canN, canS, canE, canW): for each direction, the set of source
    cells from which that one-step move is allowed (edges + walls)."""
    blockN = blockS = blockE = blockW = 0
    for (a, b) in state.h_walls:
        blockN |= _bit(a, b) | _bit(a + 1, b)            # N from (a,b),(a+1,b)
        blockS |= _bit(a, b + 1) | _bit(a + 1, b + 1)    # S from (a,b+1),(a+1,b+1)
    for (a, b) in state.v_walls:
        blockE |= _bit(a, b) | _bit(a, b + 1)            # E from (a,b),(a,b+1)
        blockW |= _bit(a + 1, b) | _bit(a + 1, b + 1)    # W from (a+1,b),(a+1,b+1)
    canN = FULL & ~_EDGE_N & ~blockN
    canS = FULL & ~_EDGE_S & ~blockS
    canE = FULL & ~_EDGE_E & ~blockE
    canW = FULL & ~_EDGE_W & ~blockW
    return canN, canS, canE, canW


def _expand(frontier, masks):
    canN, canS, canE, canW = masks
    n = (frontier & canN) << N
    s = (frontier & canS) >> N
    e = (frontier & canE) << 1
    w = (frontier & canW) >> 1
    return (n | s | e | w) & FULL


def bfs_dist(state, player):
    """Shortest path length (in steps) from the player's pawn to its goal row,
    ignoring the opponent. None if unreachable. Matches shortest_path_len."""
    c, r = state.pawns[player]
    goal = _GOAL_ROW_MASK[player]
    start = _bit(c, r)
    if start & goal:
        return 0
    masks = _can_move_masks(state)
    visited = start
    frontier = start
    dist = 0
    while frontier:
        nxt = _expand(frontier, masks) & ~visited
        if not nxt:
            return None
        dist += 1
        if nxt & goal:
            return dist
        visited |= nxt
        frontier = nxt
    return None


def path_exists(state, player):
    return bfs_dist(state, player) is not None

File: native/src/bitboard.rs
use crate::state::GameState;

const FULL: u128 = (1u128 << 81) - 1;

#[inline]
fn bit(c: i32, r: i32) -> u128 { 1u128 << (r * 9 + c) }

fn row_mask(r: i32) -> u128 { let mut m = 0u128; for c in 0..9 { m |= bit(c, r); } m }
fn col_mask(c: i32) -> u128 { let mut m = 0u128; for r in 0..9 { m |= bit(c, r); } m }

fn can_move_masks(s: &GameState) -> (u128, u128, u128, u128) {
    let (mut bn, mut bs, mut be, mut bw) = (0u128, 0u128, 0u128, 0u128);
    let mut hm = s.h_mask;
    while hm != 0 {
        let i = hm.trailing_zeros() as i32; hm &= hm - 1;
        let (a, b) = (i % 8, i / 8);
        bn |= bit(a, b) | bit(a + 1, b);
        bs |= bit(a, b + 1) | bit(a + 1, b + 1);
    }
    let mut vm = s.v_mask;
    while vm != 0 {
        let i = vm.trailing_zeros() as i32; vm &= vm - 1;
        let (a, b) = (i % 8, i / 8);
        be |= bit(a, b) | bit(a, b + 1);
        bw |= bit(a + 1, b) | bit(a + 1, b + 1);
    }
    let can_n = FULL & !row_mask(8) & !bn;
    let can_s = FULL & !row_mask(0) & !bs;
    let can_e = FULL & !col_mask(8) & !be;
    let can_w = FULL & !col_mask(0) & !bw;
    (can_n, can_s, can_e, can_w)
}

#[inline]
fn expand(frontier: u128, m: (u128, u128, u128, u128)) -> u128 {
    let (cn, cs, ce, cw) = m;
    let n = (frontier & cn) << 9;
    let s = (frontier & cs) >> 9;
    let e = (frontier & ce) << 1;
    let w = (frontier & cw) >> 1;
    (n | s | e | w) & FULL
}

pub fn bfs_dist(s: &GameState, player: usize) -> Option<u32> {
    let (c, r) = (s.pawns[player].0 as i32, s.pawns[player].1 as i32);
    let goal = if player == 0 { row_mask(8) } else { row_mask(0) };
    let start = bit(c, r);
    if start & goal != 0 { return Some(0); }
    let masks = can_move_masks(s);
    let mut visited = start;
    let mut frontier = start;
    let mut dist = 0u32;
    while frontier != 0 {
        let nxt = expand(frontier, masks) & !visited;
        if nxt == 0 { return None; }
        dist += 1;
        if nxt & goal != 0 { return Some(dist); }
        visited |= nxt;
        frontier = nxt;
    }
    None
}

pub fn path_exists(s: &GameState, player: usize) -> bool { bfs_dist(s, player).is_some() }
